File: orchestration/src/runtime/docker_backend.py
from __future__ import annotations

import time
from collections.abc import Callable
from pathlib import Path
from typing import Any

from python_on_whales import DockerClient
from python_on_whales.exceptions import DockerException

from src.runtime.errors import ServiceStartError, ServiceWaitError
from src.runtime.specs import ProbeResult, ProbeSpec, RosbagSpec, RuntimeHandle, ServiceSpec

DockerClientFactory = Callable[[], Any]


class DockerBackend:
    name = "docker"

    def __init__(
        self,
        *,
        client_factory: DockerClientFactory | None = None,
        compose_files: tuple[Path, ...] = (),
        compose_profiles: tuple[str, ...] = (),
        compose_project_name: str | None = None,
        compose_project_directory: Path | None = None,
        compose_env_files: tuple[Path, ...] = (),
    ) -> None:
        self._client_factory = client_factory or DockerClient
        self._compose_files = compose_files
        self._compose_profiles = compose_profiles
        self._compose_project_name = compose_project_name
        self._compose_project_directory = compose_project_directory
        self._compose_env_files = compose_env_files

    def _client(self) -> Any:
        if self._compose_files:
            return self._client_factory(
                compose_files=list(self._compose_files),
                compose_profiles=list(self._compose_profiles),
                compose_project_name=self._compose_project_name,
                compose_project_directory=self._compose_project_directory,
                compose_env_files=list(self._compose_env_files),
            )
        return self._client_factory()

    def compose_up(self, *, services: tuple[str, ...], detach: bool = True, build: bool = False) -> None:
        self._client().compose.up(services=list(services), detach=detach, build=build)

    def compose_stop(self, *, services: tuple[str, ...]) -> None:
        self._client().compose.stop(services=list(services))

    def compose_logs(self, *, services: tuple[str, ...], tail: int) -> str:
        return str(self._client().compose.logs(services=list(services), tail=str(tail)))

    def compose_ps_status(self, *, services: tuple[str, ...]) -> list[dict[str, object]]:
        containers = self._client().compose.ps(services=list(services), all=True)
        statuses: list[dict[str, object]] = []
        for container in containers:
            container_config = getattr(container, "config", None)
            container_state = getattr(container, "state", None)
            statuses.append(
                {
                    "id": getattr(container, "id", ""),
                    "name": getattr(container, "name", ""),
                    "image": getattr(container_config, "image", ""),
                    "status": getattr(container_state, "status", ""),
                    "running": bool(getattr(container_state, "running", False)),
                }
            )
        return statuses

    def remove_container(self, name: str, *, force: bool = True) -> None:
        try:
            self._client().remove(name, force=force)
        except DockerException as exc:
            raise ServiceWaitError(f"docker remove failed for {name}: {exc}") from exc

    def execute(self, container_name: str, command: tuple[str, ...]) -> int:
        try:
            self._client().execute(container_name, list(command))
        except DockerException as exc:
            return exc.return_code or 1
        return 0

    def start_service(self, spec: ServiceSpec) -> RuntimeHandle:
        spec.validate_for_docker()
        try:
            result = self._client().run(
                spec.image,
                list(spec.command),
                detach=spec.detach,
                remove=spec.remove,
                name=spec.container_name,
                networks=list(spec.networks),
                volumes=[mount.as_docker_tuple() for mount in spec.volumes],
                workdir=(str(spec.cwd) if spec.cwd is not None else None),
                user=spec.user,
                envs=dict(spec.env),
            )
        except DockerException as exc:
            raise ServiceStartError(f"docker service {spec.name} failed to start: {exc}") from exc
        container_id = str(getattr(result, "id", "") or spec.container_name or spec.name)
        return RuntimeHandle(
            backend=self.name,
            service_name=spec.name,
            identifier=spec.container_name or container_id,
            container_id=container_id,
            command=spec.command,
            started_at=time.time(),
            log_path=spec.log_path,
        )

    def start_rosbag(self, spec: RosbagSpec) -> RuntimeHandle:
        command = spec.command()
        service = ServiceSpec(
            name=spec.name,
            command=command,
            image="ros:jazzy",
            env=spec.env,
            cwd=spec.cwd,
            detach=True,
            required=spec.required,
            log_path=spec.log_path,
        )
        return self.start_service(service)

    def run_probe(self, spec: ProbeSpec) -> ProbeResult:
        spec.validate_for_docker()
        try:
            output = self._client().run(
                spec.image,
                list(spec.command),
                remove=True,
                name=spec.container_name,
                networks=list(spec.networks),
                volumes=[mount.as_docker_tuple() for mount in spec.volumes],
                workdir=(str(spec.cwd) if spec.cwd is not None else None),
                envs=dict(spec.env),
            )
        except DockerException as exc:
            stdout = str(exc)
            _write_probe_log(spec.log_path, stdout, "")
            return ProbeResult(
                backend=self.name,
                name=spec.name,
                return_code=exc.return_code or 1,
                stdout=stdout,
                log_path=spec.log_path,
            )
        stdout = str(output)
        _write_probe_log(spec.log_path, stdout, "")
        return ProbeResult(backend=self.name, name=spec.name, return_code=0, stdout=stdout, log_path=spec.log_path)

    def wait(self, handle: RuntimeHandle, *, timeout_sec: float | None = None) -> int:
        del timeout_sec
        try:
            rc = self._client().wait(handle.identifier)
        except DockerException as exc:
            raise ServiceWaitError(f"docker service {handle.service_name} wait failed: {exc}") from exc
        if isinstance(rc, list):
            if not rc:
                return 0
            first = rc[0]
            return int(getattr(first, "status_code", first))
        return int(rc or 0)

    def stop(self, handle: RuntimeHandle, *, timeout_sec: float = 5.0) -> None:
        del timeout_sec
        try:
            self._client().remove(handle.identifier, force=True)
        except DockerException as exc:
            raise ServiceWaitError(f"docker service {handle.service_name} stop failed: {exc}") from exc

    def logs(self, handle: RuntimeHandle | str, *, tail: int = 400) -> str:
        identifier = handle.identifier if isinstance(handle, RuntimeHandle) else handle
        try:
            return str(self._client().logs(identifier, tail=tail))
        except DockerException as exc:
            raise ServiceWaitError(f"docker logs failed for {identifier}: {exc}") from exc


def _write_probe_log(path: Path | None, stdout: str, stderr: str) -> None:
    if path is None:
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    text = stdout
    if stderr:
        text = f"{text}\n--- stderr ---\n{stderr}"
    path.write_text(text, encoding="utf-8")

File: orchestration/sim/internal/runtime/docker_backend.go
package runtime

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path"
	"strconv"
	"time"
)

type CommandRunner interface {
	Run(ctx context.Context, command string, args ...string) (CommandResult, error)
}

type CommandResult struct {
	Stdout   string
	Stderr   string
	ExitCode int
}

type ExecRunner struct{}

func (ExecRunner) Run(ctx context.Context, command string, args ...string) (CommandResult, error) {
	cmd := exec.CommandContext(ctx, command, args...)
	stdoutStderr, err := cmd.CombinedOutput()
	exitCode := 0
	if err != nil {
		exitCode = 1
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		}
	}
	return CommandResult{Stdout: string(stdoutStderr), ExitCode: exitCode}, err
}

type DockerBackend struct {
	Runner      CommandRunner
	Now         func() time.Time
	DefaultUser string
}

func NewDockerBackend(runner CommandRunner) DockerBackend {
	if runner == nil {
		runner = ExecRunner{}
	}
	return DockerBackend{
		Runner:      runner,
		Now:         time.Now,
		DefaultUser: CurrentUserSpec(),
	}
}

func (backend DockerBackend) StartService(spec ServiceSpec) (RuntimeHandle, error) {
	if err := spec.ValidateDocker(); err != nil {
		return RuntimeHandle{}, err
	}
	spec = backend.withDefaultUser(spec)
	args, err := DockerServiceArgs(spec)
	if err != nil {
		return RuntimeHandle{}, err
	}
	if spec.ContainerName != "" {
		_, _ = backend.runner().Run(context.Background(), "docker", "rm", "-f", spec.ContainerName)
	}
	result, err := backend.runner().Run(context.Background(), "docker", args...)
	if err != nil {
		_ = writeLog(spec.LogPath, result.Stdout, result.Stderr)
		return RuntimeHandle{}, fmt.Errorf("docker service %s failed: %w", spec.Name, err)
	}
	_ = writeLog(spec.LogPath, result.Stdout, result.Stderr)
	return RuntimeHandle{
		Backend:       "docker",
		ServiceName:   spec.Name,
		Identifier:    identifier(spec.ContainerName, spec.Name),
		ContainerName: spec.ContainerName,
		Command:       append([]string{"docker"}, args...),
		StartedAt:     backend.now(),
		LogPath:       spec.LogPath,
	}, nil
}

func (backend DockerBackend) StartRosbag(spec RosbagSpec) (RuntimeHandle, error) {
	service, err := RosbagServiceSpec(spec)
	if err != nil {
		return RuntimeHandle{}, err
	}
	return backend.StartService(service)
}

func (backend DockerBackend) RunProbe(spec ProbeSpec) (ProbeResult, error) {
	if err := spec.ValidateDocker(); err != nil {
		return ProbeResult{}, err
	}
	service := backend.withDefaultUser(ProbeServiceSpec(spec))
	args, err := DockerServiceArgs(service)
	if err != nil {
		return ProbeResult{}, err
	}
	ctx := context.Background()
	if spec.TimeoutSec > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, time.Duration(spec.TimeoutSec*float64(time.Second)))
		defer cancel()
	}
	result, err := backend.runner().Run(ctx, "docker", args...)
	_ = writeLog(spec.LogPath, result.Stdout, result.Stderr)
	probe := ProbeResult{
		Backend:    "docker",
		Name:       spec.Name,
		ReturnCode: result.ExitCode,
		Stdout:     result.Stdout,
		Stderr:     result.Stderr,
		LogPath:    spec.LogPath,
	}
	if err != nil {
		return probe, fmt.Errorf("docker probe %s failed: %w", spec.Name, err)
	}
	return probe, nil
}

func (backend DockerBackend) Wait(handle RuntimeHandle) (int, error) {
	result, err := backend.runner().Run(context.Background(), "docker", "wait", handle.Identifier)
	if err != nil {
		return result.ExitCode, fmt.Errorf("docker wait %s failed: %w", handle.ServiceName, err)
	}
	if result.Stdout == "" {
		return 0, nil
	}
	code, parseErr := strconv.Atoi(firstLine(result.Stdout))
	if parseErr != nil {
		return 1, fmt.Errorf("docker wait %s returned non-integer status %q", handle.ServiceName, result.Stdout)
	}
	return code, nil
}

func (backend DockerBackend) Stop(handle RuntimeHandle) error {
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()
	_, err := backend.runner().Run(ctx, "docker", "rm", "-f", handle.Identifier)
	if err != nil {
		return fmt.Errorf("docker stop %s failed: %w", handle.ServiceName, err)
	}
	return nil
}

func (backend DockerBackend) Logs(handle RuntimeHandle, tail int) (string, error) {
	if tail <= 0 {
		tail = 400
	}
	result, err := backend.runner().Run(context.Background(), "docker", "logs", "--tail", strconv.Itoa(tail), handle.Identifier)
	if err != nil {
		return result.Stdout, fmt.Errorf("docker logs %s failed: %w", handle.ServiceName, err)
	}
	return result.Stdout, nil
}

func DockerServiceArgs(spec ServiceSpec) ([]string, error) {
	if err := spec.ValidateDocker(); err != nil {
		return nil, err
	}
	args := []string{"run"}
	if spec.Detach {
		args = append(args, "--detach")
	}
	if spec.Remove {
		args = append(args, "--rm")
	}
	if spec.ContainerName != "" {
		args = append(args, "--name", spec.ContainerName)
	}
	for _, network := range spec.Networks {
		args = append(args, "--network", network)
	}
	for _, mount := range spec.Volumes {
		value, err := mount.DockerArg()
		if err != nil {
			return nil, err
		}
		args = append(args, "--volume", value)
	}
	if spec.CWD != "" {
		args = append(args, "--workdir", spec.CWD)
	}
	if spec.User != "" {
		args = append(args, "--user", spec.User)
	}
	for key, value := range spec.Env {
		args = append(args, "--env", key+"="+value)
	}
	args = append(args, spec.Image)
	args = append(args, spec.Command...)
	return args, nil
}

func ProbeServiceSpec(spec ProbeSpec) ServiceSpec {
	return ServiceSpec{
		Name:          spec.Name,
		Command:       spec.Command,
		Image:         spec.Image,
		ContainerName: spec.ContainerName,
		Env:           spec.Env,
		CWD:           spec.CWD,
		Volumes:       spec.Volumes,
		Networks:      spec.Networks,
		Detach:        false,
		Remove:        true,
		Required:      spec.Required,
		LogPath:       spec.LogPath,
		ServiceRole:   spec.ServiceRole,
	}
}

func CurrentUserSpec() string {
	uid := os.Getuid()
	gid := os.Getgid()
	if uid == 0 {
		return ""
	}
	return fmt.Sprintf("%d:%d", uid, gid)
}

func (backend DockerBackend) withDefaultUser(spec ServiceSpec) ServiceSpec {
	if spec.User != "" {
		return spec
	}
	if backend.DefaultUser == "" {
		return spec
	}
	spec.User = backend.DefaultUser
	spec.Env = withWritableRuntimeEnv(spec.Env)
	return spec
}

func withWritableRuntimeEnv(env map[string]string) map[string]string {
	if env == nil {
		env = map[string]string{}
	} else {
		copied := make(map[string]string, len(env)+3)
		for key, value := range env {
			copied[key] = value
		}
		env = copied
	}
	if env["HOME"] == "" {
		env["HOME"] = "/tmp"
	}
	if env["ROS_LOG_DIR"] == "" {
		env["ROS_LOG_DIR"] = "/tmp/navlab-ros-logs"
	}
	if env["XDG_CACHE_HOME"] == "" {
		env["XDG_CACHE_HOME"] = "/tmp/navlab-cache"
	}
	return env
}

func RosbagServiceSpec(spec RosbagSpec) (ServiceSpec, error) {
	topics, err := spec.Topics()
	if err != nil {
		return ServiceSpec{}, err
	}
	storage := spec.Storage
	if storage == "" {
		storage = "mcap"
	}
	command := []string{"ros2", "bag", "record", "-s", storage, "--compression-mode", "file", "--compression-format", "zstd", "-o", spec.OutputPath, "--topics"}
	command = append(command, topics...)
	if spec.DurationSec > 0 {
		outputParent := path.Dir(spec.OutputPath)
		metadataPath := path.Join(spec.OutputPath, "metadata.yaml")
		shellCommand := fmt.Sprintf(
			"rm -rf %s && mkdir -p %s && set +e; timeout --signal=INT %.1f %s; rc=$?; set -e; if [ \"$rc\" != \"0\" ] && [ \"$rc\" != \"124\" ] && [ \"$rc\" != \"130\" ]; then exit \"$rc\"; fi; for i in $(seq 1 40); do [ -f %s ] && exit 0; sleep 0.25; done; exit 2",
			shellQuote(spec.OutputPath),
			shellQuote(outputParent),
			spec.DurationSec,
			shellJoin(command),
			shellQuote(metadataPath),
		)
		command = []string{"bash", "-lc", shellCommand}
	}
	return ServiceSpec{
		Name:          spec.Name,
		Command:       command,
		Image:         spec.Image,
		ContainerName: spec.ContainerName,
		Env:           spec.Env,
		CWD:           spec.CWD,
		Volumes:       spec.Volumes,
		Networks:      spec.Networks,
		Detach:        true,
		Remove:        false,
		Required:      spec.Required,
		LogPath:       spec.LogPath,
		ServiceRole:   spec.ServiceRole,
	}, nil
}

func (backend DockerBackend) runner() CommandRunner {
	if backend.Runner == nil {
		return ExecRunner{}
	}
	return backend.Runner
}

func (backend DockerBackend) now() time.Time {
	if backend.Now != nil {
		return backend.Now()
	}
	return time.Now()
}

func identifier(containerName string, fallback string) string {
	if containerName != "" {
		return containerName
	}
	return fallback
}

func firstLine(value string) string {
	for index, char := range value {
		if char == '\n' || char == '\r' {
			return value[:index]
		}
	}
	return value
}

func shellJoin(args []string) string {
	quoted := make([]string, 0, len(args))
	for _, arg := range args {
		quoted = append(quoted, shellQuote(arg))
	}
	return joinWithSpace(quoted)
}

func shellQuote(value string) string {
	if value == "" {
		return "''"
	}
	return "'" + stringsReplaceAll(value, "'", "'\"'\"'") + "'"
}

func stringsReplaceAll(value string, old string, next string) string {
	if old == "" {
		return value
	}
	result := ""
	for {
		index := stringsIndex(value, old)
		if index < 0 {
			return result + value
		}
		result += value[:index] + next
		value = value[index+len(old):]
	}
}

func stringsIndex(value string, substr string) int {
	for i := 0; i+len(substr) <= len(value); i++ {
		if value[i:i+len(substr)] == substr {
			return i
		}
	}
	return -1
}

func joinWithSpace(values []string) string {
	if len(values) == 0 {
		return ""
	}
	result := values[0]
	for _, value := range values[1:] {
		result += " " + value
	}
	return result
}
